keps/sig-api-machinery/2309-idl/README.md

```diff
@@ -0,0 +1,636 @@
+
+2. **Be easy to consume by different pieces of tooling**: a serialized
+   form (think proto descriptor files) should exist that’s easy to consume
+   in the language of your choice.  This way, different tools/languages
+   can do code generation, etc from their native environments
+
+3. **Be a good code generation input**: it should be relatively
+   straightforward to generate idiomatic types in Go.  While it’s
+   a non-goal to actually implement output for other languages, it should
+   be possible to do so and get idiomatic types (consider this "external
+   work" from goal 5).
+
+4. **Be a good OpenAPI generation input**: it should be relatively
+   straightforward to compile our IDL into Kubernetes OpenAPI.
+
+5. **Enable future/external work & experimentation**: there are many
+   possible follow-up improvements or experiments that can be done with an
+   IDL. While implementing these explicitly does not fall under the
+   concrete success criteria for this proposal, the IDL pipeline should be
+   flexible enough to support developing them as future work by us or
+   external tooling by third-party projects.
+
+## Non-Goals
+
+1. **Be a custom serialization format**: unlike proto, we’re not
+   interested in *directly* describing a serialization format.  Instead,
+   we’re interested in being able to output types.go, proto IDL, code
+   suitable for serializing in JSON, etc
+
+2. **Support all features of OpenAPI**: the Kubernetes API guidelines
+   suggest avoiding certain OpenAPI constructs, while the structural
+   schema requirements explicitly reject others.  We’re interested in an
+   IDL that guides people down the right path, and that doesn’t support
+   constructs that aren’t allowed
+
+3. **Deprecate/replace OpenAPI everywhere**: it’s not currently intended
+   that this replace all uses of OpenAPI -- OpenAPI is still part of the
+   CRD specification, will still be published by the API server, etc.
+
+4. **Change k/k’s Go type representation**: the existing codebase is
+   large and intricate.  For the sake of compatibility, initial efforts
+   will focus around generating 1-1 compatible Go with the code that
+   exists currently.
+
+## Proposal
+
+***Author's Note***: *Please* read [Alternatives](#alternatives) before
+commenting "why not" or "have you considered" here.
+
+### Syntax
+
+See [the attached proposal][syntax] for grammar, syntax design, and
+in-depth examples.
+
+[syntax]: ./syntax.md
+
+### Output Scope
+
+The intial scope of this proposal is to generate two concrete artifacts:
+**CRDs** (i.e. what we have in controller-tools today) and **Go types**
+(i.e. generate types.go).
+
+The format is intended to be flexible enough that we can eventually port
+all of the generators in code-generator to make use of the IDL.  These are
+considered follow-up work after the MVP.
+
+### Concrete KubeBuilder & k/k Changes
+
+* **All types in k/k are rewritten to use KDL**.  All Go types in k/k are
```

Comment:
Implications of this change would be that we couldn't make KDL authoritative until the IDL is at least a superset of solving all problems backwards compatible k/k types have (there should be fewer than the past, but for instance you need to handle types like intstr correctly if you will generate the go types).  This is the same problem we have in gogo-protobuf today - the protobuf IDL has to be a superset of all k/k problems, and then the generation of protobuf code has to be a superset of k/k problems, and then the library itself must be maintainable and supportable (which gogo has struggled with).  But at least protobuf IDL is able to hold all of the complexity in the Go types.

Would be good to work this as a concrete example of the edge cases in k/k types - probably behavior around intorstr and empty lists in rbac.

Reply:
> there should be fewer than the past, but for instance you need to handle types like intstr correctly if you will generate the go types

I've tried to capture all the exceptions that I know about, but I'd be happy to learn if I missed some.  I consider types like intorstring to be primitives -- they have specific mappings to go types and special behavior in serialized forms (JSON, proto, etc).  This is how they're implemented in the IDL compiler.